modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/exp/IgniteSqlFunctions.java

```diff
@@ -64,46 +73,42 @@ public static String toString(BigDecimal x) {
         return x == null ? null : x.toPlainString();
     }
 
-    /** */
-    private static BigDecimal setScale(int precision, int scale, BigDecimal decimal) {
-        return precision == IgniteTypeSystem.INSTANCE.getDefaultPrecision(SqlTypeName.DECIMAL)
-            ? decimal : decimal.setScale(scale, RoundingMode.HALF_UP);
-    }
-
     /** CAST(DOUBLE AS DECIMAL). */
     public static BigDecimal toBigDecimal(double val, int precision, int scale) {
-        BigDecimal decimal = BigDecimal.valueOf(val);
-        return setScale(precision, scale, decimal);
+        return removeDefaultScale(precision, scale, toBigDecimal(BigDecimal.valueOf(val), precision, scale));
     }
 
     /** CAST(FLOAT AS DECIMAL). */
     public static BigDecimal toBigDecimal(float val, int precision, int scale) {
-        BigDecimal decimal = new BigDecimal(String.valueOf(val));
-        return setScale(precision, scale, decimal);
+        return removeDefaultScale(precision, scale, toBigDecimal(BigDecimal.valueOf(val), precision, scale));
+    }
+
+    /** Removes redundant scale in case of default DECIMAL (without passed precision and scale). */
+    private static BigDecimal removeDefaultScale(int precision, int scale, BigDecimal val) {
+        if (precision == DFLT_NUM_PRECISION && scale == 0 && val.compareTo(val.setScale(0, NUMERIC_ROUNDING_MODE)) == 0)
+            return val.setScale(0, NUMERIC_ROUNDING_MODE);
```

Comment:
`val.setScale(0, NUMERIC_ROUNDING_MODE)` executed twice, we can cache first value. Can we somehow define that BigDecimal has a scale but fraction is 0?
Do we need this method at all? Can you provide examples when double/float is converted to BigDecimal without fraction but with scale? I've commented this method, but tests are passed.

Reply:
Without this, the tests fail: `testDecimalScale()`, `testNumericConversion()`. Actually, there are questions about current precisions of numeric. But in this ticket I didn't fix anithing about it.